### ISSUE TITLE
Models underfit on highly imbalanced dataset

### ISSUE THREAD

**Issue by jomariya23156** [MalongTech/research-ms-loss#31]
I trained models with the same loss setting as mentioned in the paper; alpha=2 and beta=50. It seemed like the models can't produce good enough embedding features for the minority class (judging from the visualization with t-SNE), but they do obviously a better job for the majority class which led to poor classification results. I'd like to get some advice on how to adjust the hyperparameters or mining setting of this ms loss to better handle the highly imbalanced dataset (say having class 0 10x more samples than class 1). For additional details, I used the embedding size of 512 and the batch size of 8 (maximum capacity of my GPU because the image size is quite large).

Thanks in advance.

**Comment by bnu-wangxun** [comment]
Sorry for replying late. 
MS loss is designed to learn discriminative embeddings, However, it has no design to address the class imbalance issue. Your problem maybe cannot be solved by adjust hyper-parameters of MS loss.  To have better performance,  I advise you to try enlarge your batch-size with memory tricks[1][2], as your batch-size is quite small now.  It may helps a lot as batch-size is extremely important for the effectiveness of embedding learning, as studied in [1].
[1] Cross-Batch Memory for Embedding Learning  
[2] Momentum Contrast for Unsupervised Visual Representation Learning

**Comment by jomariya23156** [comment]
I see. Thank you for your reply. 
Maybe a bit off-topic, but from your experience and studies, how do we usually handle a class imbalance in training deep metric learning models, apart from increasing the batch size? or do you have experience with some loss functions that addressed this problem? If yes, please recommend me some of them. I kind of seriously need it right now.

**Comment by bnu-wangxun** [comment]
Class-balanced sampling or other tricks for the class imbalance issue may help. If you want design a new loss or module to solve the class imbalance specially for DML, it is an unexplored topic to the best of my knowledge. You ought to refer to the works of long tail classification area.

**Comment by jomariya23156** [comment]
That helps a lot! Thank you :)
